### PR TITLE
Direct download

### PR DESCRIPTION
**Pull request by DeepDiver1975**
requires https://github.com/owncloud/core/pull/28250

### CODE REVIEW

File: lib/s3storage.php [hunk 1]
```diff
@@ -312,7 +312,8 @@ public function getDirectDownload($urn, $versionId = null) {
 		$cmd = $this->connection->getCommand('GetObject', [
 			'Bucket'    => $this->getBucket(),
 			'Key'       => $urn,
-            'VersionId' => $versionId
+			'VersionId' => $versionId,
+			'ResponseContentType' => 'application/octet-stream'
```
Comment (butonic) [inline lib/s3storage.php]: maybe also add 'ResponseContentDisposition' => 'attachment; filename=<filename>',  ... problem is we don't have the filename here

**Comment by mrow4a** [comment]
Should I rebase and review? @DeepDiver1975 What is the schedule and priority for having it? @felixboehm

**Comment by DeepDiver1975** [comment]
Yes review please. Thx

**Comment by DeepDiver1975** [comment]
> #12 /var/www/html/core/lib/private/Files/ObjectStore/ObjectStoreStorage.php(533): OCA\\Files_Primary_S3\\S3Storage->getDirectDownload(NULL, NULL, 'Bildschirmfoto ...') #13 /var/www/html/core/lib/private/Files/Storage/Wrapper/Wrapper.php(513): OC\\Files\\ObjectStore\\ObjectStoreStorage->getDirectDownload('Bildschirmfoto ...', NULL)

looks like the file name could not be resolved to the file id ...

File: appinfo/info.xml [hunk 2]
```diff
@@ -17,6 +18,7 @@
 	</types>
 	<commands>
 		<command>OCA\Files_Primary_S3\Command\ls</command>
+		<command>\OCA\Files_Primary_S3\Command\createBucket</command>
```
Comment (butonic) [inline appinfo/info.xml]: get rid of leading slash

Comment (butonic) [reply appinfo/info.xml]: use upcase classnames `Ls`  and `CreateBucket` please

File: lib/Command/createBucket.php [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+use Aws\S3\S3Client;
+use OCP\IConfig;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class createBucket extends Command {
```
Comment (butonic) [inline lib/Command/createBucket.php]: use upcase classnames `CreateBucket` please

File: lib/s3storage.php [hunk 4]
```diff
@@ -283,4 +288,27 @@ public function saveVersion($internalPath) {
 		// In a versioned bucket the versions are created automatically
 		return true;
 	}
+
+	/**
+	 * @param string $urn
+	 * @param string $versionId
+	 * @param string $fileName
+	 * @return string mixed
```
Comment (butonic) [inline lib/s3storage.php]: string or mixed?

File: lib/s3storage.php [hunk 5]
```diff
@@ -61,16 +63,28 @@ public function __construct($params) {
 		}
 
 		$this->params = $params;
-		$this->params['autocreate'] = isset($this->params['autocreate']) ? $this->params['autocreate'] : false;
 
+		// to expensive here .... need to find a better place
```
Comment (butonic) [inline lib/s3storage.php]: add `TODO`

File: lib/Command/createBucket.php [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+			if ($client->getApi()->hasOperation('waitUntilBucketExists')) {
+				$client->waitUntil('BucketExists', [
+					'Bucket' => $bucketName,
+					'waiter.interval' => 1,
+					'waiter.max_attempts' => 15
+				]);
+			}
+		}
+
+		// enabled versioning on the bucket
```
Comment (butonic) [inline lib/Command/createBucket.php]: redundant with the log message. kill it

**Comment by butonic** [comment]
the namespace should be `FilesPrimaryS3` imo ... underscores are no longer needed. only the app id needs to match the folder name. but that is a separate PR

**Comment by DeepDiver1975** [comment]
doesn't work -> bomb

**Pull request closed without merging**